core/src/main/scala-3/toml/derivation/DerivedProductCodec.scala

```diff
@@ -0,0 +1,123 @@
+    Left(Nil, s"Cannot resolve `${label}`")
+
+  given [P <: Product](using
+    labelled: Labelling[P],
+    inst: K0.ProductInstances[Codec, P],
+    d: DefaultParams[P]
+  ): DerivedProductCodec[Option[P]] with
+    type Result[A] = Either[Parse.Error, Option[A]]
+    override def optional: Boolean = true
+    def apply(value: Value, __ : Defaults, ___ : Int): Result[P] =
+      val labels = labelled.elemLabels.iterator.zipWithIndex
+
+      val decodeField =
+        [t] => (codec: Codec[t]) =>
+          value match
+            case Value.Tbl(map) =>
+              val (witnessName, _) = labels.next()
+              map.get(witnessName) match
+                case Some(value) =>
+                  codec(value, d.defaultParams, 0)
+                    .map(Some(_))
+                    .left.map((a,m) => (witnessName +: a, m))
+                case None =>
+                  Right(
+                    d.defaultParams.get(witnessName)
+                      .map(_.asInstanceOf[t])
+                  )
+            case value =>
+              Left((Nil, "Not Implemented"))
+      val combineFields: Ap[[a] =>> Result[a]] =
+        [a, b] =>
+          (ff: Result[a => b], fa: Result[a]) =>
+            (fa, ff) match
+              case (Left(e), Right(_)) => Left(e)
+              case (_, Left(e)) => Left(e)
+              case (Right(Some(a)), Right(Some(f))) => Right(Some(f(a)))
+              case (Right(_), Right(_)) => Right(None)
+
+      inst.constructA[Result](decodeField)(
+        pure = [a] => (x: a) => Right(Some(x)),
+        map = [a, b] => (fa: Result[a], f: a => b) => fa.map:
+          case Some(a) => Some(f(a))
+          case None => None,
+        combineFields
+      )
+
+
+  given [P <: Product](using
+    labelled: Labelling[P],
+    inst: K0.ProductInstances[Codec, P],
+    d: DefaultParams[P],
+  ): DerivedProductCodec[P] with
+    override def apply(value: Value, __ : Defaults, ___ :Index): Either[Parse.Error, P] =
+      val labels = labelled.elemLabels.iterator.zipWithIndex
+      val labelsSet = labelled.elemLabels.toSet
+
+      def validateNoExtraField(map: Map[String, Value]) =
+        map.keySet.diff(labelsSet).headOption match
+          case None => Right(())
+          case Some(unknownField) =>
+            Left((List(unknownField), "Unknown field"))
+
+      val decodeField =
+        [t] => (codec: Codec[t]) =>
+          value match
+              case Value.Tbl(map) =>
+                for
+                  _ <- validateNoExtraField(map)
+                  (witnessName, _) = labels.next()
+                  result <- map.get(witnessName) match
+                    case Some(value) => codec.apply(value, d.defaultParams, 0)
+                      .left.map((a,m) => (witnessName +: a, m))
+                    case None =>
+                      d.defaultParams.get(witnessName) match
+                        case None if codec.optional => Right(None.asInstanceOf[t])
```

Comment:
optional  returns `true` if and only if codec is of type `Codec[Option[?]]`, so it is safe to use `None.asInstanceOf[t]` where `t` is of type `Option[?]`.